sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```
@@ -0,0 +1,225 @@
+     * The result will be result of currently running initialization function.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function is successfully completed it will return the value. It is possible this
+     * value is stale and will only be updated after the force refresh task is complete.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function failed - new one will be launched.
+     * </p>
+     *
+     * @param key Key for which to get a value.
+     * @param singleValueInitFunc Initialization function.
+     * @param forceRefresh Force refresh for refreshing the cache
+     * @return Cached value or value returned by initialization function.
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        boolean forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            logger.debug("cache[{}] exists", key);
+            return initialLazyValue.getValueAsync().flatMap(value -> {
+                if(!forceRefresh) {
+                    return Mono.just(value);
+                }
+
+                Mono<TValue> refreshMono = initialLazyValue.createAndWaitForBackgroundRefreshTaskAsync(singleValueInitFunc);
+                AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(refreshMono);
+                this.values.put(key, asyncLazyWithRefresh);
+                AsyncLazyWithRefresh<TValue> result = this.values.get(key);
+
+                return result.getValueAsync().onErrorResume(
+                    (error) -> {
+                        logger.debug("refresh cache [{}] resulted in error", key, error);
+                        if (initialLazyValue.shouldRemoveFromCache()) {
+                            this.remove(key);
+                        }
+                        return Mono.empty();
```

same question here: should we throw the exception instead of return empty result?